telegraf.go

```diff
@@ -60,318 +63,213 @@ type TelegrafConfigFilter struct {
+	tcd := new(telegrafConfigDecode)
 
-// telegrafConfigEncode is the helper struct for json encoding.
-type telegrafConfigEncode struct {
-	ID          *ID    `json:"id"`
-	OrgID       *ID    `json:"orgID,omitempty"`
-	Name        string `json:"name"`
-	Description string `json:"description"`
+	if err := json.Unmarshal(b, tcd); err != nil {
+		return err
+	}
 
-	Agent TelegrafAgentConfig `json:"agent"`
+	orgID := tcd.OrgID
+	if orgID == nil || !orgID.Valid() {
+		orgID = tcd.OrganizationID
+	}
 
-	Plugins []telegrafPluginEncode `json:"plugins"`
-}
+	if tcd.ID != nil {
+		tc.ID = *tcd.ID
+	}
 
-// telegrafPluginEncode is the helper struct for json encoding.
-type telegrafPluginEncode struct {
-	// Name of the telegraf plugin, exp "docker"
-	Name    string         `json:"name"`
-	Type    plugins.Type   `json:"type"`
-	Comment string         `json:"comment,omitempty"`
-	Config  plugins.Config `json:"config,omitempty"`
-}
+	if orgID != nil {
+		tc.OrgID = *orgID
+	}
 
-// telegrafConfigDecode is the helper struct for json decoding.
-type telegrafConfigDecode struct {
-	ID             *ID    `json:"id,omitempty"`
-	OrganizationID *ID    `json:"organizationID,omitempty"`
-	OrgID          *ID    `json:"orgID,omitempty"`
-	Name           string `json:"name"`
-	Description    string `json:"description"`
+	tc.Name = tcd.Name
+	tc.Description = tcd.Description
 
-	Agent TelegrafAgentConfig `json:"agent"`
+	// Prefer new structure; use full toml config.
+	tc.Config = tcd.Config
+	tc.Metadata = tcd.Metadata
 
-	Plugins []telegrafPluginDecode `json:"plugins"`
-}
+	if tcd.Plugins != nil {
+		// legacy, remove after some moons. or a migration.
+		if len(*tcd.Plugins) > 0 {
+			bkts, conf, err := decodePluginRaw(tcd)
+			if err != nil {
+				return err
+			}
+			tc.Config = plugins.AgentConfig + conf
+			tc.Metadata = map[string]interface{}{"buckets": bkts}
+		} else if c, ok := plugins.GetPlugin("output", "influxdb_v2"); ok {
+			// Handles legacy adding of default plugins (agent and output).
+			tc.Config = plugins.AgentConfig + c.Config
+			tc.Metadata = map[string]interface{}{
+				"buckets": []string{},
+			}
+		}
+	} else if tcd.Metadata == nil || len(tcd.Metadata) == 0 {
+		// Get buckets from the config.
+		m, err := parseMetadata(tc.Config)
+		if err != nil {
+			return err
+		}
 
-// telegrafPluginDecode is the helper struct for json decoding.
-type telegrafPluginDecode struct {
-	// Name of the telegraf plugin, exp "docker"
-	Name    string          `json:"name"`
-	Type    plugins.Type    `json:"type"`
-	Comment string          `json:"comment,omitempty"`
-	Config  json.RawMessage `json:"config,omitempty"`
-}
+		tc.Metadata = m
+	}
 
-// TelegrafPlugin is the general wrapper of the telegraf plugin config
-type TelegrafPlugin struct {
-	Comment string         `json:"comment,omitempty"`
-	Config  plugins.Config `json:"config,omitempty"`
-}
+	if tc.Config == "" {
+		return errors.New("no config provided")
```

Comment:
Might be good to use an `influxdb.Error` here. Typically any error that isn't an influxdb error ends up giving the user a really awful message in it's API response.